Wizard Onboarding Redesign

Sonictherocketman opened the pull request:
### Summary of Changes

The Wizard has been entirely redone and is now much smoother and quicker.

As a bonus the wizard now supports the 'Roll 4d6 and Drop the Lowest' stats rolling method.

The new wizard is much less code and far easier to understand. Also it's modern-sh JS now.

I've also removed the old dropbox and legacy import and cleaned up the design of the wizard a ton.

### Issues Fixed

N/A

### Changes Proposed (if any)

N/A

### Screen Shot of Proposed Changes (if UI related)

#### Character Creation
(excuse this picture, the formatting is off because of the crazy zoom out I had to do to get the whole form in one pic)
<img width="732" alt="Screenshot 2023-08-21 at 12 10 21 PM" src="https://github.com/adventurerscodex/adventurerscodex/assets/3310280/cad4f4cf-6079-44b8-8a9f-ef58e2704053">

#### Campaign Creation
<img width="1218" alt="Screenshot 2023-08-21 at 12 11 11 PM" src="https://github.com/adventurerscodex/adventurerscodex/assets/3310280/38828731-7e60-47d1-a650-d53515f6eebc">

#### Point Buy
<img width="1260" alt="Screenshot 2023-08-21 at 12 11 53 PM" src="https://github.com/adventurerscodex/adventurerscodex/assets/3310280/2c6cc84f-4686-406c-acfa-1c80c5ef1862">

#### Roll 4d6
<img width="1237" alt="Screenshot 2023-08-21 at 12 12 22 PM" src="https://github.com/adventurerscodex/adventurerscodex/assets/3310280/f6291926-bdaf-440d-a5e6-a220a0383270">

#### Manual Entry
<img width="1222" alt="Screenshot 2023-08-21 at 12 12 40 PM" src="https://github.com/adventurerscodex/adventurerscodex/assets/3310280/bb39dc64-e9e2-436e-9868-10b176e4388c">

tanyxp commented on src/charactersheet/viewmodels/common/wizard/4d6-drop-1/index.js:
```diff
@@ -0,0 +1,37 @@
+        this.rolls = ko.observableArray([]);
+    }
+
+    roll() {
```
very fun! cant believe we didnt have this before

tanyxp commented on src/charactersheet/viewmodels/common/wizard/index.html:
```diff
@@ -1,56 +1,335 @@
+          type="number"
+          class="form-control"
+          id="age"
+          data-bind="value: gender"
```
```suggestion
          data-bind="value: age"
```

tanyxp commented on src/charactersheet/viewmodels/common/wizard/index.js:
```diff
@@ -1,377 +1,152 @@
+                    abilityScores: ko.mapping.toJS(this.scores()),
+                    background: {
+                        name: this.background(),
+                        flaw: '',
```
do these fields have to be populated with `''`?

Sonictherocketman replied:
I believe the API requires them, yes. This is how it was before. I just kept the prior data format.

tanyxp replied:
sounds good. just couldnt remember

tanyxp commented on src/charactersheet/viewmodels/common/wizard/index.js:
```diff
@@ -1,377 +1,152 @@
+            ));
+
+            const { data } = await Hypnos.client.action({
+                keys: ['core', 'characters', 'create'],
+                params: {
+                    profile: {
+                        characterName: this.characterName(),
+                        level: this.level(),
+                    },
+                    abilityScores: ko.mapping.toJS(this.scores()),
+                    background: {
+                        name: this.background(),
+                        flaw: '',
+                        bond: '',
+                        ideal: '',
+                        personalityTrait: '',
+                    },
+                    profileImage: { type: 'email' },
+                    health: { maxHitPoints: 10 },
+                    traits: traits,
+                    items: items,
```
ah, I can see why `age, exp, race, gender, deity, class, alignment` isn't populated. it's not being set here.

tanyxp commented on src/charactersheet/viewmodels/common/wizard/index.js:
```diff
@@ -1,377 +1,152 @@
-            const items = profileData.items;
-            if (items) {
-                params.items = items;
+        this.currentState = ko.observable(State.START);
```
is this ever used?

tanyxp commented on src/charactersheet/viewmodels/common/wizard/index.js:
```diff
@@ -1,377 +1,152 @@
-    self.previousSteps = ko.observableArray([]);
-    self.currentStep = ko.observable(null);
-    self.nextStep = ko.observable(null);
+const State = {
```
is this used?

tanyxp commented on src/charactersheet/viewmodels/common/wizard/index.js:
```diff
@@ -1,377 +1,152 @@
+    }
+
+    async save() {
+        if (this.playerType() == 'character') {
```
for readability, the `if/else` can be split out into their own methods.

tanyxp reviewed (changes requested):
- age, exp, race, gender, deity, class, alignment doesnt transfer over from wizard form
- player name missing? (not character name)
- asterisk by the starting backpack, but not actually required?
- gender and age field are `data-bind`'d to the same value
- in the "roll 4d6" tab, when you hit reset, the previous roll numbers are still shown. should they also go away with the reset?
- after character creation, i get error in console for `errors.js:2 Uncaught (in promise) ParameterError: Unknown parameter: "isContainer"`
- open the console, after creating a DM, create an encounter, click through the encounter sections. clicking on some of the sections always throws an error in the logs. no idea why.

I see in the description that the import was removed. if we were to remove it, there seems to be a lot of leftover text for "import" throughout the pages. not to mention the website, reddit, and other places. part of me is `-1` on removing the import, but I could probably be convinced to be `+0/1`.

tanyxp commented on src/charactersheet/viewmodels/common/wizard/index.html:
```diff
@@ -1,56 +1,335 @@
+          type="number"
+          class="form-control"
+          id="xp"
+          data-bind="value: xp"
```
isnt this field called experience in the `.js` class?

Sonictherocketman commented:
> * age, exp, race, gender, deity, class, alignment doesnt transfer over from wizard form
Fixed
> * player name missing? (not character name)
Fixed
> * asterisk by the starting backpack, but not actually required?
Fixed
> * gender and age field are `data-bind`'d to the same value
Fixed
> * in the "roll 4d6" tab, when you hit reset, the previous roll numbers are still shown. should they also go away with the reset?
> * after character creation, i get error in console for `errors.js:2 Uncaught (in promise) ParameterError: Unknown parameter: "isContainer"`
I think this is because your API may not have the container fixes. Please update your API to ensure this is still a bug.
> * open the console, after creating a DM, create an encounter, click through the encounter sections. clicking on some of the sections always throws an error in the logs. no idea why.
No idea what this is. I will look into it.
> I see in the description that the import was removed. if we were to remove it, there seems to be a lot of leftover text for "import" throughout the pages. not to mention the website, reddit, and other places. part of me is `-1` on removing the import, but I could probably be convinced to be `+0/1`.
My justification for this was that the import is basically useless to anyone but very legacy users. I checked our analytics and we've imported 6 characters/campaigns this year and ~10 in all of 2022. TBH I don't think it's worth it to support this use-case any longer and if we want import/export then we should re-engineer the whole process for a better product.

We do have to sift through the website to find the references, but that can indeed be done.
